### PR TITLE
Memory usage limited for outgoing eth responses

### PR DESCRIPTION
**Pull request by zilm13**
@mkalinin please, take a look

### CODE REVIEW

**Comment by zilm13** [comment]
@mkalinin ready for review and merge

**Comment by zilm13** [comment]
https://github.com/ethereum/ethereumj/pull/1057/commits/a3535b004800ebb630ae7a96e2846a1f059dc033#diff-c66ab184b22490e4ae62cc3f62b9b234R44 is a bit non-effective with memory, but we could enhance it later. Usually I see not a big number of pending txs on Mainnet, so slicing is not really used.

File: ethereumj-core/src/main/java/org/ethereum/core/BlockHeader.java [hunk 1]
```diff
@@ -423,4 +425,23 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Arrays.hashCode(getHash());
     }
+
+    public static final MemSizeEstimator<BlockHeader> MemEstimator = bh ->
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/core/BlockHeader.java]: Seems unused. Let's remove it

File: ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth62.java [hunk 2]
```diff
@@ -453,8 +454,15 @@ protected synchronized void processBlockHeaders(BlockHeadersMessage msg) {
     }
 
     protected synchronized void processGetBlockBodies(GetBlockBodiesMessage msg) {
-        List<byte[]> bodies = blockchain.getListOfBodiesByHashes(msg.getBlockHashes());
-
+        Iterator<byte[]> bodiesIterator = blockchain.getIteratorOfBodiesByHashes(msg.getBlockHashes());
+        List<byte[]> bodies = new ArrayList<>();
+        int sizeSum = 80; // ArrayList skeleton
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth62.java]: i) it's a magic number ii) does `80` bytes matter much with regard to `32Kb` limit?

File: ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth62.java [hunk 3]
```diff
@@ -410,14 +412,13 @@ protected synchronized void processTransactions(TransactionsMessage msg) {
...
-        List<BlockHeader> headers = blockchain.getListOfHeadersStartFrom(
+        Iterator<BlockHeader> headersIterator = blockchain.getIteratorOfHeadersStartFrom(
                 msg.getBlockIdentifier(),
                 msg.getSkipBlocks(),
                 min(msg.getMaxHeaders(), MAX_HASHES_TO_SEND),
                 msg.isReverse()
         );
-
-        BlockHeadersMessage response = new BlockHeadersMessage(headers);
+        BlockHeadersMessage response = new BlockHeadersMessage(headersIterator);
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth62.java]: Why do we need to change `BlockHeadersMessage`?

File: ethereumj-core/src/main/java/org/ethereum/net/server/Channel.java [hunk 4]
```diff
@@ -372,8 +375,28 @@ public void prohibitTransactionProcessing() {
...
+     * Using {@link #sendTransactionsSafely(List)} is recommended instead
+     * @param txs   Transactions
+     */
+    public void sendTransactions(List<Transaction> txs) {
+        eth.sendTransaction(txs);
+    }
+
+    /**
+     * Sames as {@link #sendTransactions(List)} but input list is randomly sliced to
+     * contain not more than {@link MAX_SAFE_TXS} if needed
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/net/server/Channel.java]: `{@link MAX_SAFE_TXS}` -> `{@link #MAX_SAFE_TXS}`

File: ethereumj-core/src/main/java/org/ethereum/net/server/Channel.java [hunk 5]
```diff
@@ -372,8 +375,28 @@ public void prohibitTransactionProcessing() {
...
+     * @param txs   List of txs to send
+     */
+    public void sendTransactionsSafely(List<Transaction> txs) {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/net/server/Channel.java]: `Safely` sounds too abstract for me and confusing on the other hand (is regular `sendTransactions` unsafe??)

File: ethereumj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 6]
```diff
@@ -511,11 +513,16 @@ public String toString(int maxDataSize) {
         this.rlpEncoded = RLP.encodeList(nonce, gasPrice, gasLimit,
                 receiveAddress, value, data, v, r, s);
 
-        this.hash = this.getHash();
+        this.hash = HashUtil.sha3(rlpEncoded);
 
         return rlpEncoded;
     }
 
+    public synchronized void purgeEncoded() {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/core/Transaction.java]: it's not used anywhere except tests. I'd remove it therefore

**Comment by zilm13** [comment]
@mkalinin done!

File: ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth63.java [hunk 7]
```diff
@@ -130,17 +130,21 @@ protected synchronized void processGetReceipts(GetReceiptsMessage msg) {
         );
 
         List<List<TransactionReceipt>> receipts = new ArrayList<>();
+        int sizeSum = 0;
         for (byte[] blockHash : msg.getBlockHashes()) {
             Block block = blockchain.getBlockByHash(blockHash);
             if (block == null) continue;
 
             List<TransactionReceipt> blockReceipts = new ArrayList<>();
+            sizeSum += 0;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/net/eth/handler/Eth63.java]: 😀

File: ethereumj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 8]
```diff
@@ -252,10 +255,9 @@ public boolean isParsed() {
 
     public byte[] getHash() {
         if (!isEmpty(hash)) return hash;
-
         rlpParse();
-        byte[] plainMsg = this.getEncoded();
-        return HashUtil.sha3(plainMsg);
+        getEncoded();
+        return hash;
```
Comment (ESchouten) [inline ethereumj-core/src/main/java/org/ethereum/core/Transaction.java]: This change causes an empty array to be returned when parsed == true && rlpEncoded != null. Happends all the time with the StandaloneBlockchain.

Please revert this method to:

`public byte[] getHash() {
        if (!isEmpty(hash)) return hash;
         rlpParse();
        byte[] plainMsg = this.getEncoded();
        return HashUtil.sha3(plainMsg);
}`

**Pull request merged**